trio/_core/tests/test_io.py

```diff
@@ -273,3 +273,65 @@ def fileno_wrapper(fileobj):
 
     assert results["send_a"] == results["recv_b"]
     assert results["send_b"] == results["recv_a"]
+
+
+async def test_notify_closing_on_invalid_object():
+    # It should either be a no-op (generally on Unix, where we don't know
+    # which fds are valid), or an OSError (on Windows, where we currently only
+    # support sockets, so we have to do some validation to figure out whether
+    # it's a socket or a regular handle).
+    got_oserror = False
+    got_no_error = False
+    try:
+        trio.hazmat.notify_closing(-1)
+    except OSError:
+        got_oserror = True
+    else:
+        got_no_error = True
+    assert got_oserror or got_no_error
+
+
+async def test_io_manager_statistics():
+    def check(stats, expected_readers, expected_writers):
+        if stats.backend in ["epoll", "windows"]:
+            assert stats.tasks_waiting_read == expected_readers
+            assert stats.tasks_waiting_write == expected_writers
+        else:
+            stats.backend == "kqueue"
+            assert stats.tasks_waiting == expected_readers + expected_writers
+
+    a1, b1 = stdlib_socket.socketpair()
+    a2, b2 = stdlib_socket.socketpair()
+    a3, b3 = stdlib_socket.socketpair()
+    for sock in [a1, b1, a2, b2, a3, b3]:
+        sock.setblocking(False)
+    with a1, b1, a2, b2, a3, b3:
+        # let the call_soon_task settle down
+        await wait_all_tasks_blocked()
+
+        statistics = _core.current_statistics()
+        print(statistics)
+        # 1 for call_soon_task
+        check(statistics.io_statistics, 1, 0)
+
+        # We want:
+        # - one socket with a writer blocked
+        # - two sockets with a reader blocked
+        # - a socket with both blocked
+        fill_socket(a1)
+        fill_socket(a3)
+        async with _core.open_nursery() as nursery:
+            nursery.start_soon(_core.wait_writable, a1)
+            nursery.start_soon(_core.wait_readable, a2)
+            nursery.start_soon(_core.wait_readable, b2)
+            nursery.start_soon(_core.wait_writable, a3)
+            nursery.start_soon(_core.wait_readable, a3)
+
+            await wait_all_tasks_blocked()
+
+            statistics = _core.current_statistics()
+            print(statistics)
+            # 1 for call_soon_task
+            check(statistics.io_statistics, 4, 2)
+
+            nursery.cancel_scope.cancel()
```

Comment:
`check(_core.current_statistics().io_statistics, 1, 0)` again after the nursery block?

I also wonder if it's worth having a version of this test that wakes up the waiters normally (write data on the other side to wake the readers, read from the other side to wake the writers) rather than cancelling them, so we get coverage on both paths.

Reply:
I don't think we need to fret too much over this test; it's just a quick check that some rarely-used almost-trivial debug code works properly. If things are so broken that the IOManager can't track cancellation correctly, then hopefully we have other tests that will notice that :-)

Reply:
I did do a cleanup pass and add the check at the end though.